cmd/clusterctl/client/cluster/components.go

```diff
@@ -168,19 +169,30 @@ func (p *providerComponents) Delete(options DeleteOptions) error {
 			namespacesToDelete.Insert(obj.GetName())
 		}
 
+		// If the resource is part of the inventory for clusterctl don't delete it at this point as losing this information makes the
+		// upgrade function non-reentrant. Instead keep the inventory objects around until the upgrade is finished and working and
+		// delete them at the end of the upgrade flow.
+		isInventory := obj.GroupVersionKind().GroupKind().String() == providerGroupKind
+		if isInventory {
+			if !options.IncludeInventory {
+				continue
+			}
+		}
+
 		// If the resource is a cluster resource, skip it if the resource name does not start with the instance prefix.
 		// This is required because there are cluster resources like e.g. ClusterRoles and ClusterRoleBinding, which are instance specific;
 		// During the installation, clusterctl adds the instance namespace prefix to such resources (see fixRBAC), and so we can rely
 		// on that for deleting only the global resources belonging the the instance we are processing.
 		// NOTE: namespace and CRD are special case managed above; webhook instead goes hand by hand with the controller they
 		// should always be deleted.
 		isWebhook := obj.GroupVersionKind().Kind == validatingWebhookConfigurationKind || obj.GroupVersionKind().Kind == mutatingWebhookConfigurationKind
+
 		if util.IsClusterResource(obj.GetKind()) &&
 			!isNamespace && !isCRD && !isWebhook &&
+			!isInventory &&
```

Comment:
I don't this this test is required here, given that inventory resource are namespaced and they will be filtered out by `util.IsClusterResource(obj.GetKind())` at the beginning of the if statement

Reply:
That's right - removed